pkg/admissioncontroller/webhooks/auth/seed/authorizer.go

```diff
@@ -73,75 +73,68 @@ func (a *authorizer) Authorize(_ context.Context, attrs auth.Attributes) (auth.D
+				[]string{"create"},
+				nil,
+			)
 		}
 	}
 
 	return auth.DecisionNoOpinion, "", nil
 }
 
-func (a *authorizer) authorizeGet(seedName string, fromType graph.VertexType, attrs auth.Attributes) (auth.Decision, string, error) {
-	if ok, reason := a.checkVerb(seedName, attrs, "get"); !ok {
-		return auth.DecisionNoOpinion, reason, nil
-	}
-
-	if ok, reason := a.checkSubresource(seedName, attrs); !ok {
-		return auth.DecisionNoOpinion, reason, nil
-	}
-
-	return a.authorize(seedName, fromType, attrs)
-}
-
-func (a *authorizer) authorizeBackupBucket(seedName string, attrs auth.Attributes) (auth.Decision, string, error) {
-	if ok, reason := a.checkVerb(seedName, attrs, "get", "list", "watch", "create", "update", "patch", "delete"); !ok {
-		return auth.DecisionNoOpinion, reason, nil
-	}
-
-	if ok, reason := a.checkSubresource(seedName, attrs, "status"); !ok {
-		return auth.DecisionNoOpinion, reason, nil
-	}
-
-	// When a new BackupBucket is created then it doesn't yet exist in the graph, so we only handle update operations
-	// here. The create case is handled in the SeedRestriction admission handler.
-	switch attrs.GetVerb() {
-	case "create", "list", "get", "watch":
-		return auth.DecisionAllow, "", nil
-	default:
-		return a.authorize(seedName, graph.VertexTypeBackupBucket, attrs)
-	}
+func (a *authorizer) authorizeRead(seedName string, fromType graph.VertexType, attrs auth.Attributes) (auth.Decision, string, error) {
+	return a.authorize(seedName, fromType, attrs, []string{"get"}, nil, nil)
 }
 
-func (a *authorizer) authorizeShootState(seedName string, attrs auth.Attributes) (auth.Decision, string, error) {
-	if ok, reason := a.checkVerb(seedName, attrs, "get", "create", "update", "patch"); !ok {
+func (a *authorizer) authorize(
+	seedName string,
+	fromType graph.VertexType,
+	attrs auth.Attributes,
+	allowedVerbs []string,
+	alwaysAllowedVerbs []string,
+	allowedSubresources []string,
+) (
+	auth.Decision,
+	string,
+	error,
+) {
+	if ok, reason := a.checkVerb(seedName, attrs, allowedVerbs...); !ok {
 		return auth.DecisionNoOpinion, reason, nil
 	}
 
-	if ok, reason := a.checkSubresource(seedName, attrs); !ok {
+	if ok, reason := a.checkSubresource(seedName, attrs, allowedSubresources...); !ok {
 		return auth.DecisionNoOpinion, reason, nil
 	}
 
-	// When a new ShootState is created then it doesn't yet exist in the graph, so we only handle update operations
-	// here. The create case is handled in the SeedRestriction admission handler.
-	if attrs.GetVerb() == "create" {
+	// When a new object is created then it doesn't yet exist in the graph, so usually such requests are always allowed
+	// as the 'create case' is typically handled in the SeedRestriction admission handler. Similarly, resources for
+	// which the gardenlet has a controller need to be listed/watched, so those verbs would also be allowed here.
+	if utils.ValueExists(attrs.GetVerb(), alwaysAllowedVerbs) {
 		return auth.DecisionAllow, "", nil
 	}
```

Comment:
I wonder if we should move this check to the top so that `allowedVerbs` doesn't have to contain all elements of  `alwaysAllowedVerbs`. WDYT?

Concretely we can rearrange the order to:
- `checkSubresource`
- `utils.ValueExists(attrs.GetVerb(), alwaysAllowedVerbs)`
- `a.checkVerb(seedName, attrs, allowedVerbs...)`

Reply:
Yes, good idea, thanks! I've done it with 8d87acc.

PS: When calling the `checkVerb` function I merge `allowedVerbs` and `alwaysAllowedVerbs` to provide a more meaningful error message in case this function returns an error.